ethereum/rpc/namespaces/debug/api.go

```diff
@@ -112,6 +112,85 @@ func (a *API) TraceTransaction(hash common.Hash, config *evmtypes.TraceConfig) (
 	return decodedResult, nil
 }
 
+func (a *API) TraceBlockByNumber(height rpctypes.BlockNumber, config *evmtypes.TraceConfig) ([]*TxTraceResult, error) {
+	a.logger.Debug("debug_traceBlockByNumber", "height", height)
+	if height == 0 {
+		return nil, errors.New("genesis is not traceable")
+	}
+	// Get Tendermint Block
+	resBlock, err := a.backend.GetTendermintBlock(height)
+	if err != nil {
+		return nil, err
+	}
+
+	txsLength := len(resBlock.Block.Txs)
+
+	if txsLength == 0 {
+		// If there are no transactions return empty array
+		return []*TxTraceResult{}, nil
+	}
+
+	var (
+		results = make([]*TxTraceResult, txsLength)
+		wg = new(sync.WaitGroup)
+		jobs = make(chan *TxTraceTask, txsLength)
+	)
+
+	threads := runtime.NumCPU()
+	if threads > txsLength {
+		threads = txsLength
+	}
+	wg.Add(threads)
+	for th := 0; th < threads; th++ {
+		go func() {
+			defer wg.Done()
+			// Fetch and execute the next transaction trace tasks
+			for task := range jobs {
+				tx, err := a.clientCtx.TxConfig.TxDecoder()(resBlock.Block.Txs[task.Index])
+				if err != nil {
+					a.logger.Debug("tx not found", "hash", resBlock.Block.Txs[task.Index].Hash())
+					continue
+				}
+
+				ethMessage, ok := tx.GetMsgs()[0].(*evmtypes.MsgEthereumTx)
+				if !ok {
+					// Just considers Ethereum transactions
+					continue
+				}
+				traceTxRequest := &evmtypes.QueryTraceTxRequest{
+					Msg:     ethMessage,
+					TxIndex: uint32(task.Index),
```

Comment:
is this cast secure? i.e no overflow?

Reply:
yeah this could overflow, better check at the beginning of the function call if `txsLength` overflows uint32